sgkit/stats/ld.py

```diff
@@ -0,0 +1,60 @@
+import math
+
+import numba
+import numpy as np
+from numba import njit
+
+from sgkit.typing import ArrayLike
+
+
+# Note that float32 types are used in this function to get an exact match with scikit-allel's equivalent function.
+@njit(nogil=True, fastmath=True, cache=True, locals={"m0": numba.float32, "m1": numba.float32, "v0": numba.float32, "v1": numba.float32, "cov": numba.float32, "n": numba.int32})  # type: ignore
+def rogers_huff_r_between(gn0: ArrayLike, gn1: ArrayLike) -> float:  # pragma: no cover
+    """Rogers Huff R
+
+    From https://github.com/cggh/scikit-allel/blob/961254bd583572eed7f9bd01060e53a8648e620c/allel/opt/stats.pyx
+    """
+    # initialise variables
+    m0 = m1 = v0 = v1 = cov = 0.0
+    n = 0
+
+    # iterate over input vectors
+    for i in range(len(gn0)):
+        x = gn0[i]
+        y = gn1[i]
+        # consider negative values as missing
+        if x >= 0 and y >= 0:
+            n += 1
+            m0 += x
+            m1 += y
+            v0 += x ** 2
+            v1 += y ** 2
+            cov += x * y
+
+    # early out
+    if n == 0:
+        return np.nan
+
+    # compute mean, variance, covariance
+    m0 /= n
+    m1 /= n
+    v0 /= n
+    v1 /= n
+    cov /= n
+    cov -= m0 * m1
+    v0 -= m0 * m0
+    v1 -= m1 * m1
+
+    d = math.sqrt(v0 * v1)
+    if d < np.finfo(np.float32).tiny:
```

Comment:
I believe my reason for adding this was that without it, you can end up with results outside of [0, 1].  This may not fix that in all cases though.  We may want to consider instead returning `nan` if `d` is too small (like this) but also clip `r` to [0, 1].  It seems odd to let `r` be either in [0, 1] or `inf` but I'm good with that too.  Perhaps there's a better way?

Reply:
A lot of the time it's simpler and more consistent to just let the floating point error happen and IEEE define what the exact semantics should be. What happens if we just don't make the check here? I think it's fine that r is in [0, 1] except for cases where we hit against numerical limits (and use the IEEE semantics for those limits)

Reply:
That works for me.  I'd like to suggest that we then add a note in the docs that inf is a possibility to avoid users tripping up on something like `xr.DataArray([np.inf]).isnull().item(0) == False # or np.isnan(np.inf)`, but I'm sure we've not documented it in other places.  I'd still be weakly in favor of documenting it anyways since I know I would appreciate it as a user.

Reply:
Sounds good @eric-czech, good for us to document and test these corner cases. In general, I guess we should encourage people to use ``np.isfinite`` rather than ``not np.isnan`` to make client code robust, though.

Reply:
👍 

I wish Xarray had made that as prominent as `DataArray.isnull` rather than having to do `DataArray.pipe(xr.ufuncs.isfinite)`.

Reply:
@eric-czech file an issue upstream?

Reply:
https://github.com/pydata/xarray/issues/5261

We should talk to Quansight about getting a bot to file Xarray issues like this automatically 😉

Reply:
+1 to removing the if clause. If we also replace the division as follows

```
-    r = cov / d
+    r = np.divide(cov, d)
```

Then it would return `np.inf` in some cases. However, I have not been able to find what these cases are, so I'm not sure how to document them. (An input containing `np.inf` will return `np.nan` for example, since when computing the covariance `np.inf - np.inf` is calculated, which is `np.nan`.)

So we could just say that users should be prepared to handle `np.nan` and infinities, or we could ensure that the return value is always a finite float or `np.nan` by adding:

```
+    if np.isinf(r):
+        return np.nan
```

Thoughts?

Reply:
I think we can probably just say that the method will return non-finite values (IEEE inf and NaNs) under certain situations (e.g AN EXAMPLE), and the user should be prepared to deal with that by checking np.isfinite. Checking isfinite isn't any worse than check isnan.

The trouble with putting in special cases for numerical corner cases is that you nearly always miss some. Simpler in the long run to just let the maths and IEEE semantics do the work.

I don't feel that strongly about it though, feel free to ignore this.